### PR TITLE
Metadata action type

### PR DESCRIPTION
**Pull request by rahulanand16nov**
#10 

Implements metadata acton type along with some unit tests.

I have not done the end-to-end test yet because changes are required in the Kuadrant controller before we test the user-id-based rate limit.

### CODE REVIEW

File: src/filter/http_context.rs [hunk 1]
```diff
@@ -228,7 +228,36 @@ impl Filter {
...
+                        None => {
+                            debug!("metadata key not found");
+                            descriptor_entry.set_key(default_value.into());
+                        }
+                        Some(metadata_bytes) => match String::from_utf8(metadata_bytes) {
+                            Err(e) => {
+                                debug!("failed to parse metadata value: {}", e);
+                                descriptor_entry.set_value(default_value.into());
+                            }
+                            Ok(metadata_value) => descriptor_entry.set_value(metadata_value),
```
Comment (eguzki) [inline src/filter/http_context.rs:257]: `descriptor_entry` needs to be added to the `entries`
```rust
entries.push(descriptor_entry);
```

File: src/filter/http_context.rs [hunk 2]
```diff
@@ -228,7 +228,36 @@ impl Filter {
...
+                RLA_action_specifier::metadata(md) => {
+                    // Note(rahul): defaulting to dynamic metadata source right now.
+                    let metadata_key = &md.get_metadata_key().key;
+                    let mut metadata_path: Vec<&str> = md
+                        .get_metadata_key()
+                        .get_path()
+                        .iter()
+                        .map(|path_segment| path_segment.get_key())
+                        .collect();
+                    let default_value = md.get_default_value();
```
Comment (eguzki) [inline src/filter/http_context.rs:240]: default value is a required fields? in the envoy docs, it is not https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-msg-config-route-v3-ratelimit-action-metadata

Maybe it is in the protoc objects.

What if the default value is not set?

```
default_value

    ([string](https://developers.google.com/protocol-buffers/docs/proto#scalar)) An optional value to use if metadata_key is empty. If not set and no value is present under the metadata_key then no descriptor is generated.
```

**Review by eguzki — CHANGES_REQUESTED** [review]

**Comment by eguzki** [comment]
> I have not done the end-to-end test yet because changes are required in the Kuadrant controller before we test the user-id-based rate limit.

You can test locally with the docker compose env

File: src/filter/http_context.rs (hunk 2 above)
Comment (rahulanand16nov) [reply src/filter/http_context.rs:240]: It would use the default `String` value i.e. `""`

File: src/filter/http_context.rs (hunk 1 above)
Comment (rahulanand16nov) [reply src/filter/http_context.rs]: fixed

File: src/configuration.rs [hunk 3]
```diff
@@ -141,6 +163,35 @@ mod test {
...
+        let configurations = &gateway_actions[0].configurations;
+        assert_eq!(configurations.len(), 1);
+
+        let actions = &configurations[0].actions;
+        assert_eq!(actions.len(), 2);
+        assert!(std::matches!(
+            actions[0],
+            RLA_action_specifier::generic_key(_)
+        ));
+        assert!(std::matches!(actions[1], RLA_action_specifier::metadata(_)));
```
[the remark this answers is not in the document]
Comment (rahulanand16nov) [reply src/configuration.rs]: fixed

File: deploy/Makefile [hunk 4]
```diff
@@ -2,15 +2,16 @@ MKFILE_PATH := $(abspath $(lastword $(MAKEFILE_LIST)))
 PROJECT_PATH := $(patsubst %/,%,$(dir $(MKFILE_PATH)))
 IMAGE_VERSION ?= latest
 DOCKER ?= podman
+IMAGE_REPOSITORY ?= rahanand
```
[the remark this answers is not in the document]
Comment (rahulanand16nov) [reply deploy/Makefile]: fixed

File: src/filter/http_context.rs (hunk 2 above)
Comment (rahulanand16nov) [reply src/filter/http_context.rs:240]: Looks like it will fail the parsing of configuration and yes, it's due to protoc objects.

**Comment by rahulanand16nov** [comment]
> > I have not done the end-to-end test yet because changes are required in the Kuadrant controller before we test the user-id-based rate limit.
> 
> You can test locally with the docker compose env. No need for the kuadrant's env to test it.

I need some way to set the dynamic metadata which is done by Authorino. :thinking:

File: src/filter/http_context.rs (hunk 2 above)
Comment (rahulanand16nov) [reply src/filter/http_context.rs:240]: I wonder if `default_value` is empty then we should skip this action altogether.

**Comment by rahulanand16nov** [comment]
@eguzki I have used the `header-to-metadata` filter and updated the local development configuration to use metadata action. LGTM.

File: src/filter/http_context.rs (hunk 2 above)
Comment (eguzki) [reply src/filter/http_context.rs:240]: I would say so

Comment (rahulanand16nov) [reply src/filter/http_context.rs:240]: I think a better option is to let RLP validation tell the user that the default value is not provided.

Comment (rahulanand16nov) [reply src/filter/http_context.rs:240]: Done.

**Comment by rahulanand16nov** [comment]
@eguzki Okay to merge now?

**Review by eguzki — APPROVED** [review]

**Pull request merged**